GBPTree should not checkpoint during close if panic

burqen opened the pull request:
Checkpoint will increment stable and unstable generation of tree, making it impossible to recognise crash-pointers during the recovery following the database crash.

Commit:
Extract interface from DatabaseHealth

and move interface to neo4j-common so that modules below neo4j-kernel
can depend on it.

Commit:
GBPTree dont checkpoint on close if panic

Commit:
Fix checkstyle

Commit:
Fix bug, native label scan store missing entry 0,0

When searching for allNodeLabelRanges, native label scan store
would miss entry with labelId 0 and nodeId 0 if it was the only
entry because it did not calculate highestLabelId correctly.

Commit:
Add possibility to register monitor for native label scan store

And underlying GBPTree through Monitors.

Commit:
Integration test, should not checkpoint GBPTree if panic

Commit:
Cleanup GBPTreeTest

Commit:
Fix compilation fix after rebase

Commit:
Fix unsatisfied dependencies for LabelScanStore

MishaDemianenko commented on community/common/src/main/java/org/neo4j/kernel/Health.java:
```diff
@@ -0,0 +1,56 @@
+ */
+package org.neo4j.kernel;
+
+public interface Health
```
Lets add javadocs for this shiny new interface

MishaDemianenko commented on community/common/src/main/java/org/neo4j/kernel/Health.java:
```diff
@@ -0,0 +1,56 @@
+     * @param panicDisguise the cause of the unhealthy state wrapped in an exception of this type.
+     * @throws EXCEPTION exception type to wrap cause in.
+     */
+    default <EXCEPTION extends Throwable> void assertHealthy( Class<EXCEPTION> panicDisguise ) throws EXCEPTION
```
Why all methods in this newly introduced interface are default?

burqen replied:
To simplify `Adapter` implementation. Will remove `Adapter` and create more descriptive class as suggested 👍

MishaDemianenko commented on community/common/src/main/java/org/neo4j/kernel/Health.java:
```diff
@@ -0,0 +1,56 @@
+        return null;
+    }
+
+    class Adapter implements Health
```
What if instead of this simple Adapter class we will have Health implementor in it's own class file with descriptive name?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -757,9 +765,12 @@ private void close( Header.Writer headerWriter ) throws IOException
-            // Perform a checkpoint before closing. If no changes has happened since last checkpoint,
-            // no new checkpoint will be created.
-            checkpoint( IOLimiter.unlimited(), headerWriter );
+            if ( health.isHealthy() )
```
Maybe lets add some log statement for case when we skip checkpointing?

burqen replied:
Ye, it would be good. It should to be logged from `LabelScanStore` level though as `GBPTree` is just some underlying implementation detail.

It is not super clear to me how this should be with relation to `LuceneLabelScanStore`.

But maybe something like:
- Add `close( didCheckpoint )` to `GBPTree.Monitor` and trigger during `GBPTree.close()`.
- Propagate monitor call to `LabelScanStore.Monitor` which register it as `shutdown( didCheckpoint )`. 
- `LoggingMonitor` implementation log some line about this.

Problem: What does `LabelScanStore.shutdown( didCheckpoint )` mean for `Lucene` implementation?

tinwelint replied:
I like the monitor approach. If `didCheckpoint` would instead be called e.g. `cleanShutdown` it would apply better to other label index implementations too.

burqen replied:
👍

burqen replied:
@tinwelint , I talked to @MishaDemianenko about it. He suggested just injecting a `Log` instance and use that directly.

Please have a look at commit `2217c3b` to see what you think about it.

tinwelint replied:
@burqen @MishaDemianenko I can't agree with that. May I say that Monitor is _always_ preferred over direct Log. A Monitor has way more usability, is customizable and is easier to test (you do want to test this don't you). We don't use direct logging anywhere else in index component, why here?

MishaDemianenko replied:
@tinwelint can you accept current way of logging, or you would prefer monitor instead? I'm fine with any of those, sure monitor have much wider range of applicability but since here we just need to log i think using just log is fine (and we have assertable log to be able to test that)

tinwelint replied:
@MishaDemianenko It's just a very strange inconsistency in the index component, to bypass the usual pattern for a particular event. Also, it's much better to have the layer outside gb+tree, e.g. label scan store to log something more contextual, saying that "native label scan store didn't checkpoint blabla" instead of having a log statement saying that some gb+tree did something... what's that to the user? Weird I'd say. Think about logging as surface too.

Let's keep abstractions and try not to bypass for quick fixes. Please.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java:
```diff
@@ -107,6 +114,10 @@
     private final FullStoreChangeStream fullStoreChangeStream;
 
     /**
+     * todo
```
todo

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/internal/BatchInserterImpl.java:
```diff
@@ -278,7 +280,9 @@ public BatchInserterImpl( final File storeDir, final FileSystemAbstraction fileS
         Dependencies deps = new Dependencies();
-        deps.satisfyDependencies( fileSystem, config, logService, indexStoreView, pageCache );
+        Monitors monitors = new Monitors();
+        Health databaseHealth = new Health.Adapter();
```
do we really need to use dummy health here?

burqen replied:
We need _some_ `Health` to make the `KernelExtension` load and we don't need full `DatabaseHealth` so might as well use most simple one. No?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/store/BatchingNeoStores.java:
```diff
@@ -161,6 +162,8 @@ private BatchingNeoStores( FileSystemAbstraction fileSystem, PageCache pageCache
         dependencies.satisfyDependency( IndexStoreView.EMPTY );
         dependencies.satisfyDependency( pageCache );
+        dependencies.satisfyDependency( new Monitors() );
+        dependencies.satisfyDependency( new Health.Adapter() );
```
same here, do we need to use dummy health here?

burqen replied:
Same as above

MishaDemianenko reviewed (changes requested):
Couple of comments, good otherwise

Commit:
Javadoc for Health

Commit:
Introduce AlwaysHealthy

Commit:
Introduce abstract CountingMonitor

to simplify implementation of counting monitors in GBPTree

Commit:
Missing javadoc for NativeLabelScanStore

Commit:
Log checkpoint action during GBPTree.close

MishaDemianenko commented on community/common/src/main/java/org/neo4j/kernel/AlwaysHealthy.java:
```diff
@@ -0,0 +1,24 @@
+ */
+package org.neo4j.kernel;
+
+public class AlwaysHealthy extends Health.Adapter
```
Now we have `AlwaysHealthy` and Adapter and first one which is basically Adapter except it's public, what if we will have SilentHealth (current AlwaysHealthy thingy) and will remove Adapter since all implementations can extend that SilentHealth instead. WDYT?

burqen replied:
Sounds good, I was not very happy with the solution I came up with but I wanted the flexibility of an Adapter and also the "more useful" class name.
`SilentHealth` seems like a good middle road 👍

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -754,16 +772,25 @@ public void close( Consumer<PageCursor> headerWriter ) throws IOException
         finally
         {
             pagedFile.close();
+            String checkpointMessage = didCheckpoint ?
```
i would just say: `Skip index checkpoint since database is not healthy.` and did not log anything in any other case, but maybe that just me so if you think that strange or not enough, just ignore this comment.

burqen replied:
I think will keep it since the problem we are addressing is when we do checkpoints even though we shouldn't.

tinwelint replied:
See above. We can't really know what to call this thing here. `index`, meh, for different uses this is a different thing, e.g. in label scan store this is a `label index`, for a schema index this is a `schema index for label L and property P`, not just `index`. This has to be a monitor to be done properly.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/index/GBPTreeLogTest.java:
```diff
@@ -0,0 +1,128 @@
+
+    private final Layout<MutableLong,MutableLong> layout = new SimpleLongLayout();
+
+    private void insert( GBPTree<MutableLong, MutableLong> index, long key ) throws IOException
```
please move private helper methods bellow all the tests

MishaDemianenko reviewed (commented):
Couple of minor comments/questions

Commit:
Rename AlwaysHealthy to SilentHealth and remove Health.Adapter

Commit:
Move helper methods to bottom in tests

burqen commented:
@MishaDemianenko Comments addressed and PR rebased

MishaDemianenko commented:
@burqen fine for me, lets wait for @tinwelint if he is fine with proposed changes.

Commit:
GBPTree never checkpoint on close

NativeLabelScanStore take over responsibility for checkpointing
underlying GBPTree. It will do so only if, at shutdown, no panic
has occurred and NativeLabelScanStore was successfully started,
meaning any recovery was completed.

When recovering the tree we rely on the last successful checkpoint
before the crash that made recovery necessary. If recovery fails we
will retry it again on next start up. If the tree was checkpointed
then the next recovery will not be able to reliably clean up all
'crashed' pointers.

burqen commented:
@MishaDemianenko @tinwelint Some new commits to this bad boy.
Whole responsibility for checkpointing `GBPTree` moves to `NativeLabelScanStore`. This is to avoid unsafe checkpoints in the case where recovery fails.

Commit:
Remove unnecessary tests

Commit:
Cleanup GBPTreeTest

Commit:
Missing license header

tinwelint commented:
@burqen With the recent developments the DatabaseHealth refactorings are a bit unnecessary, aren't they Although I'd guess it doesn't harm us. One thing you could do though is to reduce `Health` down to `isHealthy()` w/ perhaps additionally `assertHealthy()` because I'd like to have the database, i.e. for now the specific implementation `DatabaseHealth` to control when panic happens.

So simply reduce number of methods in `Health` interface.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java:
```diff
@@ -426,6 +443,12 @@ public void stop() throws IOException
+        monitor.flushDuringShutdown( flush );
+        if ( flush )
+        {
+            index.checkpoint( IOLimiter.unlimited() );
```
shouldn't we close index first (at least index writer) and then checkpoint?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStore.java:
```diff
@@ -426,6 +443,12 @@ public void stop() throws IOException
     public void shutdown() throws IOException
     {
+        boolean flush = started && health.isHealthy();
+        monitor.flushDuringShutdown( flush );
```
why do we call this thing flush instead of checkpoint? wouldn't that be a bit confusing?

MishaDemianenko commented:
@burqen 2 comments left :)

burqen commented:
This PR is getting out of hand. Redo fix in https://github.com/neo4j/neo4j/pull/9111
Will cherry-pick the good stuff from this PR into new smaller PRs.
